internal/store/db/mongo/store.go

```diff
@@ -16,37 +16,254 @@
 package mongo
 
 import (
+	"context"
+	"errors"
+	"fmt"
+	"reflect"
+	"strconv"
+	"time"
+
+	"github.com/edgexfoundry/app-functions-sdk-go/internal/store/contracts"
 	"github.com/edgexfoundry/app-functions-sdk-go/internal/store/db"
 	"github.com/edgexfoundry/app-functions-sdk-go/internal/store/db/interfaces"
-	"github.com/edgexfoundry/app-functions-sdk-go/internal/store/models"
+	"github.com/edgexfoundry/app-functions-sdk-go/internal/store/db/mongo/models"
+
+	"go.mongodb.org/mongo-driver/bson"
+	"go.mongodb.org/mongo-driver/bson/primitive"
+	"go.mongodb.org/mongo-driver/mongo"
+	"go.mongodb.org/mongo-driver/mongo/options"
 )
 
+// Client provides a wrapper for Mongo's Client type
+type Client struct {
+	Timeout time.Duration
+	Client  *mongo.Database
+	// unexported Client for Disconnect
+	client *mongo.Client
+}
+
+const mongoCollection = "store"
+
 // Store persists a stored object to the data store.
-func Store(o models.StoredObject) error {
-	return nil
+func (c Client) Store(o contracts.StoredObject) (string, error) {
+	ctx, cancel := context.WithTimeout(context.Background(), c.Timeout)
+	defer cancel()
+
+	objID, uuid, err := models.FromContractId(o.ID)
+	if err != nil {
+		return "", err
+	}
+	var doc bson.M
+
+	if objID == primitive.NilObjectID {
+		// determine if this object already exists in the DB
+		filter := bson.M{"uuid": uuid}
+		result := c.Client.Collection(mongoCollection).FindOne(ctx, filter)
+
+		var m models.StoredObject
+		_ = result.Decode(&m)
+
+		// if the result of the lookup is any object other than the empty, it exists
+		if !reflect.DeepEqual(m, models.StoredObject{}) {
+			return "", errors.New("object exists in database")
+		}
+
+		doc = bson.M{
+			"uuid":             uuid,
+			"appServiceKey":    o.AppServiceKey,
+			"payload":          o.Payload,
+			"retryCount":       o.RetryCount,
+			"pipelinePosition": o.PipelinePosition,
+			"version":          o.Version,
+			"correlationID":    o.CorrelationID,
+			"eventID":          o.EventID,
+			"eventChecksum":    o.EventChecksum,
+		}
+	} else {
+		doc = bson.M{
+			"_id":              objID,
+			"uuid":             uuid,
+			"appServiceKey":    o.AppServiceKey,
+			"payload":          o.Payload,
+			"retryCount":       o.RetryCount,
+			"pipelinePosition": o.PipelinePosition,
+			"version":          o.Version,
+			"correlationID":    o.CorrelationID,
+			"eventID":          o.EventID,
+			"eventChecksum":    o.EventChecksum,
+		}
+	}
+
+	result, err := c.Client.Collection(mongoCollection).InsertOne(ctx, doc)
+	if err != nil {
+		return "", err
+	}
+
+	return result.InsertedID.(primitive.ObjectID).Hex(), nil
```

Comment:
What is this used for?

Reply:
This is the ID that Mongo generates with a Store operation. I thought we'd decided we wanted to generate an ID during the Store operation, and I changed the API to return that ID so that you can know what it is without having to do a second DB lookup.